core/workflow-core/src/main/scala/edu/uci/ics/amber/core/storage/result/iceberg/IcebergDocument.scala

```diff
@@ -0,0 +1,269 @@
+package edu.uci.ics.amber.core.storage.result.iceberg
+
+import edu.uci.ics.amber.core.storage.IcebergCatalogInstance
+import edu.uci.ics.amber.core.storage.model.{BufferedItemWriter, VirtualDocument}
+import edu.uci.ics.amber.core.storage.util.StorageUtil.{withLock, withReadLock, withWriteLock}
+import edu.uci.ics.amber.core.tuple.Tuple
+import edu.uci.ics.amber.util.IcebergUtil
+import org.apache.iceberg.{DataFile, FileScanTask, Table}
+import org.apache.iceberg.catalog.{Catalog, TableIdentifier}
+import org.apache.iceberg.data.{IcebergGenerics, Record}
+import org.apache.iceberg.exceptions.NoSuchTableException
+import org.apache.iceberg.io.CloseableIterable
+import org.apache.iceberg.types.Types
+
+import java.net.URI
+import java.util.concurrent.locks.{ReentrantLock, ReentrantReadWriteLock}
+import scala.jdk.CollectionConverters._
+
+/**
+  * IcebergDocument is used to read and write a set of T as an Iceberg table.
+  * It provides iterator-based read methods and supports multiple writers to write to the same table.
+  *
+  * - On construction, the table will be created if it does not exist.
+  * - If the table exists, it will be overridden.
+  *
+  * @param tableNamespace namespace of the table.
+  * @param tableName name of the table.
+  * @param tableSchema schema of the table.
+  * @param serde function to serialize T into an Iceberg Record.
+  * @param deserde function to deserialize an Iceberg Record into T.
+  * @tparam T type of the data items stored in the Iceberg table.
+  */
+class IcebergDocument[T >: Null <: AnyRef](
+    val tableNamespace: String,
+    val tableName: String,
+    val tableSchema: org.apache.iceberg.Schema,
+    val serde: (org.apache.iceberg.Schema, T) => Record,
+    val deserde: (org.apache.iceberg.Schema, Record) => T
+) extends VirtualDocument[T] {
+
+  private val lock = new ReentrantReadWriteLock()
+
+  @transient lazy val catalog: Catalog = IcebergCatalogInstance.getInstance()
+
+  // During construction, create or override the table
+  synchronized {
```

Comment:
this `synchronized` is unnecessary as it locks `this` instance.

Reply:
removed